### PR TITLE
Reduxify

### PR DESCRIPTION
**Pull request by mgerring**

### CODE REVIEW

**Comment by mgerring** [comment]
Two of the test suites in npm aren't passing yet and there may be some issues with the python tests as well, but you can start reviewing now

File: assets/js/workflow-reducer.js [hunk 1]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
 }
+registerReducerFunc(ADD_MODULE + '_FULFILLED', (state, action) => {
+  let insertBefore = action.payload.insert_before;
 
-// Call delete API, then dispatch a reload
-export function removeModuleAction(id_to_delete) {
+  delete action.payload.insert_before;
+
+  return update(state, {
+    selected_wf_module: {$set: action.payload.id},
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: Is this necessary, since you also dispatch setSelectedWfModuleAction? Maybe you can avoid that setSelectedWfModuleAction entirely, and hence the thunk.

File: assets/js/workflow-reducer.js [hunk 2]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
 
-export function changeSelectedWfModuleAction(wfModuleID) {
-  return {
-    type : SELECTED_MODULE_CHANGE,
-    id : wfModuleID,
+  return function (dispatch) {
+    return (
+      // Set the new selected module before deleting to avoid errors.
+      // We do this even if the selected module isn't changing to avoid
+      // writing a subtly tricky conditional here.
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: It does cost an extra server round-trip, however. Or maybe we do this optimistically anyway so we delete on the client without waiting?

File: assets/js/workflow-reducer.js [hunk 3]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+  let workflowID = store.getState().workflow.id;
   return {
-    type : TOGGLE_MODULE_COLLAPSED,
-    id : wfModuleID,
-    isCollapsed
+    type : SET_SELECTED_MODULE,
+    payload : {
+      promise: api.setSelectedWfModule(workflowID, wfModuleID).then((response) => {
+        return {
+          wf_module_id: response.selected_wf_module
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: Why is this needed? I would think this value only appears in _FULFILLED, and we don't bother handling it.

File: assets/js/workflow-reducer.js [hunk 4]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+        }
+      }),
+      data: {
+        wf_module_id: wfModuleID
+      }
+    }
   }
 }
+registerReducerFunc(SET_SELECTED_MODULE + '_PENDING', (state, action) => {
+  if (!('selected_wf_module' in state) || (action.payload.wf_module_id !== state.selected_wf_module)) {
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: When would it happen that state does not have selected_wf_module, if it comes in from serializer?

File: assets/js/workflow-reducer.js [hunk 5]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+// Patch a workflow module with new data
 
-            // Clone the state, switch out this one wfm (keep position in wf_modules array )
-            newState = Object.assign({}, state);
-            newState.workflow = Object.assign({}, state.workflow);
-            newState.workflow.wf_modules = state.workflow.wf_modules.map(el => el!==wfm ? el : newWfm);
-          }
+// TODO: We don't validate which fields or types are on
+// a WfModule here. The backend will reject nonexistent
+// fields, but should we do something on the frontend?
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: I'm comfortable not checking, for the following reason: the backend MUST check, as a security concern. If the front end calls this with bogus fields it's a bug, but no harm done. However it does bring up the issue of run time assertions. We currently don't have any mechanism for them, and it's not clear what we would do with assertion failures if we did.

File: assets/js/workflow-reducer.js [hunk 6]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+  };
+}
+registerReducerFunc(UPDATE_WF_MODULE + '_PENDING', (state, action) => {
+  let moduleIdx = findIdxByProp(
+    state.workflow.wf_modules,
+    'id',
+    action.payload.id
+  );
+
+  if (typeof moduleIdx !== 'undefined') {
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: Who would call with undefined?

File: assets/js/workflow-reducer.js [hunk 7]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+  return {
+    type : SET_WF_MODULE_STATUS,
+    payload: {
+      id : wfModuleID,
+      status : status,
+      error_msg: error_msg
+    }
+  }
+}
+registerReducerFunc(SET_WF_MODULE_STATUS, (state, action) => {
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: Don't we need  + "PENDING" here? Is this a bug? If so, do the tests catch it?

File: assets/js/workflow-reducer.js [hunk 8]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
+}
+registerReducerFunc(MARK_DATA_VERSIONS_READ + '_PENDING', (state, action) => {
+  let wfModuleIdx = findIdxByProp(state.workflow.wf_modules, 'id', action.payload.id);
+  if (typeof wfModuleIdx !== 'undefined' &&
+    typeof state.workflow.wf_modules[wfModuleIdx].versions !== 'undefined') {
+    return update(state, {
+      workflow: {
+        wf_modules: {
+          [wfModuleIdx]: {
+            // Take the versions array,
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: good comments

File: assets/js/workflow-reducer.js [hunk 9]
```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
...
-        return Object.assign({}, state, {
-          loggedInUser: action.user,
-        });
-      } else {
-        return state;
+export function clearNotificationsAction(wfModuleId) {
+  return {
+    type: CLEAR_NOTIFICATIONS,
+    payload: {
+      promise: api.deleteWfModuleNotifications(wfModuleId),
```
Comment (jstray) [inline assets/js/workflow-reducer.js]: I wonder if this api could be merged into updateWfModule

File: assets/js/workflow-reducer.test.js [hunk 10]
```diff
@@ -27,62 +45,216 @@ describe('Reducer actions', () => {
   };
 
   // Stub result to be returned by our stub loadWorkflow
-  const mock_load_workflow_result = { id: 1001 };
+  //const mock_load_workflow_result = { id: 1001 };
```
Comment (jstray) [inline assets/js/workflow-reducer.test.js]: Delete if un-needed

File: assets/js/workflow-reducer.test.js [hunk 11]
```diff
@@ -27,62 +45,216 @@ describe('Reducer actions', () => {
...
+  it('Adds a module', () => {
+    let state = workflowReducer(test_state, {
+      type: 'ADD_MODULE_FULFILLED',
+      payload: mock_add_wf_module_result,
+    });
+    expect(state.workflow.wf_modules[2].id).toEqual(40);
+    expect(state.selected_wf_module).toEqual(40);
+  });
+
+  it('Deletes a module', () => {
```
Comment (jstray) [inline assets/js/workflow-reducer.test.js]: Would be good to have tests for boundary conditions, so that all code paths are exercised, e.g. delete last remaining module, delete currently selected module

File: assets/js/workflow-reducer.test.js [hunk 12]
```diff
@@ -27,62 +45,216 @@ describe('Reducer actions', () => {
...
+        }
+      }
+    });
+    expect(state).toBe(test_state);
+  });
 
-    var actionPromise = removeModuleAction(30);
+  it('Sets the wfModule status', () => {
+    let state = workflowReducer(test_state, {
+      type: 'SET_WF_MODULE_STATUS',
```
Comment (jstray) [inline assets/js/workflow-reducer.test.js]: I believe this should be + '_PENDING'

File: assets/js/workflow.page.js [hunk 13]
```diff
@@ -27,16 +27,17 @@ const mapStateToProps = (state) => {
...
 const mapDispatchToProps = (dispatch) => {
   return {
     addModule: (module_id, insertBefore) => {
       dispatch(Actions.addModuleAction(module_id, insertBefore))
     },
     removeModule: (wf_module_id) => {
-      dispatch(Actions.removeModuleAction(wf_module_id))
+      dispatch(Actions.deleteModuleAction(wf_module_id))
     },
+    //TODO: make this use the redux action
```
Comment (jstray) [inline assets/js/workflow.page.js]: Why not now?

File: server/tests/test_workflow_views.py [hunk 14]
```diff
@@ -142,7 +142,10 @@ def test_workflow_addmodule_put(self):
         self.assertEqual(response.data['id'], wfm1.id)
         # we should get a full serialization back, same as /wfmodules/xx call
         serializer = WfModuleSerializer(wfm1)
-        self.assertEqual(response.data, serializer.data)
+        # addmodule serialization should also return insert_before property
+        return_data = serializer.data
+        return_data['insert_before'] = '0'
```
Comment (jstray) [inline server/tests/test_workflow_views.py]: Could return order instead

File: server/serializers.py [hunk 15]
```diff
@@ -87,11 +87,17 @@ def get_notification_count(self, wfm):
     def get_html_output(self, wfm):
         return wfm.module_version.html_output
 
+    versions = serializers.SerializerMethodField()
+    def get_versions(self, wfm):
```
Comment (jstray) [inline server/serializers.py]: This needs a test within test_wfmodule.py

File: assets/js/wfparameters/DataVersionSelect.js [hunk 16]
```diff
@@ -39,36 +39,30 @@ export default class DataVersionSelect extends React.Component {
...
-
       if (!this.state.modalOpen === false) {
-        var idx = this.state.versions.versions.map((vers) => {
-          return vers[0]
-        }).indexOf(this.state.versions.selected);
-
+        // TODO: this.props.versions.selected should be a tuple of (<Date>, <Boolean>)
+        // TODO: Don't do Array.find, it's slow
+        let idx;
+        for (let i = 0; i < this.props.versions.versions.length; i++) {
```
[the remark this answers is not in the document]
Comment (mgerring) [reply assets/js/wfparameters/DataVersionSelect.js]: array.find is indefensibly slower than a loop. A loop is something like 90% faster. I can set up the jsperf test again if you want to see for yourself.

File: assets/js/workflow-reducer.js (hunk 1 above)
Comment (mgerring) [reply assets/js/workflow-reducer.js]: It's necessary to make the UI feel fast, yes, but it's not technically necessary. Without it there's a perceptible lag.

File: assets/js/wfparameters/DataVersionSelect.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (jstray) [reply assets/js/wfparameters/DataVersionSelect.js]: Ok. Can we put our loop in a fn for clarity and DRY?

File: assets/js/workflow-reducer.js (hunk 1 above)
Comment (jstray) [reply assets/js/workflow-reducer.js]: Is there some way to dispatch both at once? I see no technical problem having two promises unresolved at the same time, and the server-side race condition is irrelevant here

File: assets/js/wfparameters/DataVersionSelect.js [hunk 17]
```diff
@@ -39,36 +39,30 @@ export default class DataVersionSelect extends React.Component {
...
+          idx = i;
+          if (this.props.versions.selected === this.props.versions.versions[i][0]) {
+            break;
+          }
+        }
         // If the selected version is false,
-        if (!this.state.versions.versions[idx][1]) {
```
[the remark this answers is not in the document]
Comment (mgerring) [reply assets/js/wfparameters/DataVersionSelect.js]: Nope, the comment still applies

File: assets/js/wfparameters/DataVersionSelect.js [hunk 18]
```diff
@@ -216,6 +196,34 @@ export default class DataVersionSelect extends React.Component {
   }
 }
 
+const mapStateToProps = (state, ownProps) => {
+  let wfModuleIdx;
+  for (let i = 0; i < state.workflow.wf_modules.length; i++) {
```
[the remark this answers is not in the document]
Comment (mgerring) [reply assets/js/wfparameters/DataVersionSelect.js]: This is arguably a huge potential performance problem since this runs every time the state updates

File: assets/js/workflow-reducer.js (hunk 1 above)
Comment (mgerring) [reply assets/js/workflow-reducer.js]: No, there isn't. You can't dispatch setSelectedWfModuleAction until you have a wfModuleId, which you only have once the wfModule returns from the database.

File: assets/js/workflow-reducer.js (hunk 4 above)
Comment (mgerring) [reply assets/js/workflow-reducer.js]: No idea, that condition was there when I started working on this

Comment (mgerring) [reply assets/js/workflow-reducer.js]: Actually, does it get returned from the serializer if it's a new workflow?

File: assets/js/workflow-reducer.js (hunk 6 above)
Comment (mgerring) [reply assets/js/workflow-reducer.js]: There's some weird cases where this could happen -- let's say a module is in some kind of queue to get updated (say, if the upstream data changed), and each update takes a long time, and before we get to this module, it gets deleted.
